xc/device.ml

```diff
@@ -1682,16 +1682,19 @@ let __start (task: Xenops_task.t) ~xs ~dmpath ?(timeout = !Xenopsd.qemu_dm_ready
 	let () = match info.disp with
 	| VNC (Vgpu [{implementation = Nvidia vgpu}], _, _, _, _)
 	| SDL (Vgpu [{implementation = Nvidia vgpu}], _) -> begin
-		(* The below line does nothing if the device is already bound to the
-		 * nvidia driver. We rely on xapi to refrain from attempting to run
-		 * a vGPU on a device which is passed through to a guest. *)
-		PCI.bind [vgpu.physical_pci_address] PCI.Nvidia;
-		let args = vgpu_args_of_nvidia domid info.vcpus vgpu in
-		let ready_path = Printf.sprintf "/local/domain/%d/vgpu-pid" domid in
-		let cancel = Cancel_utils.Vgpu domid in
-		let vgpu_pid = init_daemon ~task ~path:!Xc_path.vgpu ~args
-			~name:"vgpu" ~domid ~xs ~ready_path ~timeout:!Xenopsd.vgpu_ready_timeout ~cancel () in
-		Forkhelpers.dontwaitpid vgpu_pid
+		if not (Vgpu.is_running ~xs domid) then begin
```

Comment:
A minor idea: to avoid deeper and deeper nesting, how about pulling the `if` condition into a `when` of the pattern?

```
| SDL (Vgpu [{implementation = Nvidia vgpu}], _) when Vgpu.is_running ~xs domid ->
  D.info "Daemon %s is already running for domain %d" !Xc_path.vgpu domid
| SDL (Vgpu [{implementation = Nvidia vgpu}], _) ->
 ...
```